### PR TITLE
Add new SQUAD example

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR adds a new example for SQUAD (v1 and v2) for simple models (e.g., not the XLNet/XLM more complex version, another example will follow for those) using the datasets library and all the features of the fast tokenizer to simplify considerably the preprocessing and the post-processing.

I've compared the new version to the old one and did not find major differences when:
- fine-tuning a model on SQUAD v1 or v2 with the old and new script
- evaluation an existing model fine-tuned on SQUAD v1 or v2 with the old and new script

The only difference I found was when evaluating an existing model fine-tuned on SQUAD v1 and evaluating it on SQUAD v2. For those, the new script is a bit less good at predicting the null answers (but those models have crappy results on SQUAD v2 anyway, they are just a bit more crappy).

Further plans are:
- add a subclass of Trainer for QA so that the evaluation is done directly with `trainer.evaluate()`
- add a script for XLNet/XLM

### CODE REVIEW

**Review by LysandreJik — APPROVED** [review]
This looks good to me! This version is so much cleaner than the previous one with `squad_convert_examples_to_features`. Thank you!

**Review by LysandreJik — APPROVED** [review]
I have no preference concerning whether to use the trainer subclass or not; however, since we insist that users should do it this way, I think this example does a fantastic job at showcasing how to do it.

The version where it's at right now (after the second commit) looks great.

File: examples/question-answering/run_qa.py [hunk 1]
```diff
@@ -113,7 +113,12 @@ class DataTrainingArguments:
         default=False, metadata={"help": "If true, some of the examples do not have an answer."}
     )
     null_score_diff_threshold: float = field(
-        default=0.0, metadata={"help": "If `null_score - best_non_null` is greater than the threshold predict null."}
+        default=0.0,
+        metadata={
+            "help": "The threshold used to select the null answer: if the best answer has a score that is less than "
+            "the score of the null answer minus this threshold, the null answer is selected for this example. "
+            "Only useful when `version_2_with_negative=True`."
```
Comment (LysandreJik) [inline examples/question-answering/run_qa.py:120]: Very nice!

File: examples/question-answering/run_qa.py [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+    )
+    config_name: Optional[str] = field(
+        default=None, metadata={"help": "Pretrained config name or path if not the same as model_name"}
+    )
+    tokenizer_name: Optional[str] = field(
+        default=None, metadata={"help": "Pretrained tokenizer name or path if not the same as model_name"}
+    )
+    cache_dir: Optional[str] = field(
+        default=None,
+        metadata={"help": "Where do you want to store the pretrained models downloaded from huggingface.co"},
```
Comment (patrickvonplaten) [inline examples/question-answering/run_qa.py:65]:
```suggestion
        metadata={"help": "path to directory to store the pretrained models downloaded from huggingface.co"},
```

File: examples/question-answering/run_qa.py [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+    )
+    use_fast_tokenizer: bool = field(
+        default=True,
+        metadata={"help": "Whether to use one of the fast tokenizer (backed by the tokenizers library) or not."},
```
Comment (patrickvonplaten) [inline examples/question-answering/run_qa.py:69]:
```suggestion
        metadata={"help": "Whether to use one of the fast tokenizer (backed by the tokenizers library)."},
```

File: examples/question-answering/run_qa.py [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+    context_column_name = "context" if "context" in column_names else column_names[1]
+    answer_column_name = "answers" if "answers" in column_names else column_names[2]
+
+    # Padding side determines if we do (question|context) or (context|question).
+    pad_on_right = tokenizer.padding_side == "right"
+
+    # Training preprocessing
+    def prepare_train_features(examples):
+        # Tokenize our examples with truncation and maybe padding, but keep the overflows using a stride. This results
+        # in one example possible giving several features when a context is long, each of those features having a
```
Comment (patrickvonplaten) [inline examples/question-answering/run_qa.py:255]:
```suggestion
        # in one example with possibly several features when a context is long. Each of those features then has a
```

File: examples/question-answering/run_qa.py [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+        # The offset mappings will give us a map from token to character position in the original context. This will
+        # help us compute the start_positions and end_positions.
+        offset_mapping = tokenized_examples.pop("offset_mapping")
+
+        # Let's label those examples!
+        tokenized_examples["start_positions"] = []
+        tokenized_examples["end_positions"] = []
+
+        for i, offsets in enumerate(offset_mapping):
+            # We will label impossible answers with the index of the CLS token.
```
Comment (patrickvonplaten) [inline examples/question-answering/run_qa.py:280]: does this hold true for all models that CLS token should be used for impossible answers? What if the model does not have a cls token? Should we maybe put the "`id_for_impossible_answer`" into the model arguments and let it default to cls_token_id?

**Review by patrickvonplaten — APPROVED** [review]
Looks very clean! I'm asking myself whether we want to support scripts for qa datasets like `trivia_qa`: https://huggingface.co/datasets/viewer/?dataset=trivia_qa that don't rely on "`answer_start`" and "`answer_end`" in this `question-answering` folder as well. If yes, are the functionalities general enough to support a rather different dataset structure? Or do we just want to add a `trivia_qa.py` script in this case...

**Comment by sgugger** [comment]
I'd like the new examples scripts to stay fairly focused on one problem (at the cost of potentially have some duplicate codes) so they're easy to understand (and tweak) by users. We don't support any kind of datasets either (if your QA dataset has fields with names slightly different than SQUAD for instance), users are supposed to adapt the relevant lines in the code to their needs.

So with that in mind, I'd definitely prefer a separate script :-)

File: examples/question-answering/run_qa.py (hunk 4 above)
Comment (sgugger) [reply examples/question-answering/run_qa.py:255]: The sentence I obtain after the suggestion does not make sense to me: an example will general several inputs for the models, but it does not have them inside.

**Pull request closed without merging**